doc/rados/operations/pg-repair.rst

```diff
@@ -1,4 +1,53 @@
+============================
 Repairing PG inconsistencies
 ============================
+Sometimes a placement group might become "inconsistent". To return the
+placement group to an active+clean state, you must first determine which 
+of the placement groups has become inconsistent and then run the "pg
+repair" command on it. This page contains commands for diagnosing placement
+groups and the command for repairing placement groups that have become
+inconsistent.
+
+.. highlight:: console
+
+Commands for Diagnosing Placement-group Problems
+================================================
+The commands in this section provide various ways of diagnosing broken placement groups.
+
+The following command provides a high-level (low detail) overview of the health of the ceph cluster::
+
+  [node] # ceph health detail
+
+The following command provides more detail on the status of the placement groups::
+
+  [node] # sudo ceph pg dump --format=json-pretty
+
+The following command lists inconsistent placement groups::
+
+  [node] # rados list-inconsistent-pg PGID 
+
+The following command lists inconsistent rados objects::
+
+  [node] # rados list-inconsistent-obj PGID 
+ 
+The following command lists inconsistent snapsets in the given placement group::
+
+  [node] # rados list-inconsistent-snapset PGID
+
+
+Commands for Repairing Placement Groups
+=======================================
+The form of the command to repair a broken placement group is::
+
+  [node] # ceph pg repair PGID
+  
+Where "PGID" is the id of the affected placement group.
+
+For example::
 
+  [node] # ceph pg repair 1.4
```

Comment:
it might be worth noting what "pg repair" command actually does.

Ceph stores and updates the checksum of an object  stored in the cluster. after a scrub is performed on a PG, the osd tries to pick an authoritative copy from all its replicas. although there is only a single case of consistency, there are different forms of inconsistencies. for instance, after a deep scrub, we calculate the checksum of object read from disk, and compare it with the one previously recorded. if they don't match, it's an inconsistency. in the case of replica pool, if any of checksum the replicas of the object does not match with that of the authoritative copy, it's also considered as an occurrence of  inconsistency.

"pg repair" tries it best to fix different kinds of inconsistencies. in the first case above, if the digest of the authoritative copy exists, it tries to overwrite the inconsistent copy's digest with the one from the authoritative copy. in the second case, it simply marks the inconsistent copy missing, and leave it to recovery process to do its trick.

the same applies to the omap data.

please note, "pg repair" is not a silver bullet. Ceph does not want to be "smart" when it comes to consistency of user data. that's also why we don't repair an PG when inconsistency is found in it. because of couple reasons.

the checksum of an object or an omap item is not always available. if an object is updated in the middle, instead of recalculate the checksum by reading the whole object, we mark its checksum invalidated. in that case, even though an authoritative copy is picked when performing scrub, the level of authoritativeness of it is questionable. so Ceph will refuse to "repair" it even if "pg repair" is requested by user. and manual intervention would be necessary in this case.

an other reason is, in theory, the matched recored checksum and calculated one can not prove that the "authoritative" copy is authoritative. as there is rare chance that bitrot has changed the data on disk in a way that makes these two digests match. that's also why we want to involve human intervention when inconsistency happens.

@dzafman @joshd @neha-ojha could you help comment on this? am i missing anything?

Reply:
What is "human intervention" in the context of a discovered inconsistency? I understand that human intervention might be beyond the scope of the pg-repair page, but an explanation of the human intervention step that follows the discovery of an inconsistency should be documented somewhere.

Reply:
Also, how does Ceph determine which copy is authoritative?

Reply:
@zdover23 Ceph looks at all the replicas/shards for errors and the first one, starting with the primary, is selected as the authoritative copy.  There are many error checks that are too numerous to describe here.  Of course, for erasure coded pools data repair requires using enough copies to reconstruct the data.  The auth copy is used to repair internal state (object info, snapset) no matter the pool type.

When an authoritative copy can't be found it is because every replica/shard has 1 or more errors.  In that case something like the ceph-objectstore-tool would be used to attempt to fix at least 1 copy.  For replicated pools it should be the one with correct data content if that can be determined.  See my talk in particular the part going over the "Repair" section of the slides.

Reply:
> What is "human intervention" in the context of a discovered inconsistency? I understand that human intervention might be beyond the scope of the pg-repair page, but an explanation of the human intervention step that follows the discovery of an inconsistency should be documented somewhere.

Zack, i think "human intervention" is the whatever user needs to do to fix the inconsistencies. for instance, i think check the output of "rados list-inconsistent-pg" and issue "ceph pg repair" commands is "human intervention". the reason why i think it's important to note this is that we cannot (and probably should not) fix the inconsistency on behalf of user without using some heuristic algorithm. actually, we select the authoritative copy is based on many error checks. but even with these checks, in theory, we still cannot be 100% sure that the selected authoritative copy is correct if inconsistency is found. that's why we need "human intervention".

Reply:
tchaikov, I assumed that "human intervention" meant "whatever in the repair process is not reducible to a heuristic algorithm", and I understand that the heuristic algorithm represented by "ceph pg repair" is the best that machines can do, unaided by the case-by-case human judgment represented by "human intervention". Right now I'm at a place in my Ceph education where I don't have a feel yet for what I might do in a case requiring human intervention, and I'm trying for my own sake to imbue the phrase "human intervention" with some instructive content so that I can articulate to myself what it might be, other than "that which the command 'ceph pg repair' does not do".

I'll type up the good information that you and D. Zafman have provided, and I'll keep reaching, occasionally this publicly, for a way to write about that liminal place between algorithm and human judgment.

Reply:
>  I understand that the heuristic algorithm represented by "ceph pg repair" is the best that machines can do, unaided by the case-by-case human judgment represented by "human intervention"

yeah. but since the selection of authoritative copy is based on a heuristic algorithm, we need user to make the final call, which is a "human intervention", even it's no more than pulling a trigger, or issuing the command.

a more involved "human intervention" will be needed if Ceph are so confident about the reliability of "authoritative copy", as David put in his reply and his talk on this. in that case, user will have to use their domain knowledge to identify the corrupted or the authoritative copy. and fix it. i am afraid that it would render the document lengthy and convoluted if we want to cover this. but just for your reading fun, https://www.sebastien-han.fr/blog/2015/04/27/ceph-manually-repair-object/ is one way to fix the inconsistency.

Reply:
@tchaikov Thanks for that link. That's just what I was looking for, for my own edification.